### DISCUSSION TITLE
Modbus Server - Server hooks

### DISCUSSION THREAD

**Discussion by rwskinner** [pymodbus-dev/pymodbus#2270] [General]
Are there any types of hooks/callback or custom handler that gets called during a register read or write?
I need some custom code to perform certain actions, especially during a write for a HR.

**Comment by janiversen** [comment]
Please read the documentation and/or examples, there are several hooks available.

**Comment by rwskinner** [comment]
I've been reading the docs all day and I still don't know how to accomplish what I need.  

When a server received a write command 6/16 from a client, I want to have some options to reject it if, I don't want it writeable, or to make sure the values is within a certain range.

I'm playing with the CustomDataBlock, the the setValues handles all writes, not just the ones from a client.  IOW,  getting values from an instrument and using setValues is how I fill my registers for the client.

    def setValues(self, address, value):
        super(CustomDataBlock, self).setValues(address, value)

So, is there a way for me to do:
1) Only get the writes performed by the client.
2) Validate they fit my criteria, if so, write it, else respond with Illegal Value?
3) If the value was valid and I allow a write, then I need to publish that value to a broker to handle outside the container.

Right now, Kura Wires -Modbus asset publish mb values to a internal broker.  The container subscribes to it then when a valid mqtt message comes in, the values are used to fill the modbus server's registers.    This all works great.  Now I need to do the reverse, and publish data back when a register is written with an allowed value.

PymodbusTcp has what I'm talking about and that's why I'm trying to figure it out for pyModbus.

    def write_h_regs(self, address, words_l, srv_info):
         if some condition that is true:
            return super().write_h_regs(address, words_l, srv_info)
        else:
            return DataHandler.Return(exp_code=EXP_ILLEGAL_FUNCTION)

**Reply by janiversen** [reply]
Seems you did not read the pymodbus documentation.

Making your own datastore is one way to go, using the simulator datastore is another way to go.....and finally you could simply use the trace hooks which gives you the full freedom.

But if pymodbusTCP does what you want then stick with that.
